design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,192 @@
+modifications.
+
+## Design Details
+
+The following addition should be done **before** the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, BMO will use its own architecture as the default.
+
+- Prepare `HardwareProfile` for deprecation:
+
+  Add a new profile `empty` that does not have any fields populated. When this
+  profile is used, do not set the fields on the Ironic node that are not
+  overridden explicitly (via the `Architecture` and `RootDeviceHints`).
+
+  Remove `RootGB` (only makes sense for partition images which we don't
+  support) and `LocalGB` (optional and discovered by inspection).
+
+- Prepare `Image.ChecksumType` for the default change:
+
+  Add a new type `auto` that will tell Ironic to detect the checksum type
+  itself (by not providing the `image_os_hash_algo` field).
+
+- Ensure that a `HardwareData` CR exists for any host that has
+  `HardwareDetails` in the status.
+
+The following significant changes will be done to the new version:
+
+- Check the default value of `Image.ChecksumType` from `md5` to `auto`.
+  Mark the `md5` type as deprecated for removal before `v1`.
+
+- Remove `HardwareProfile`, effectively using the `empty` profile in the
+  new API version.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+The following cosmetic changes will be done to the new version:
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+### Implementation Details/Notes/Constraints
+
+Since the new API version will mostly involve removing or renaming fields,
+the storage version will stay `v1alpha1` until this version is removed.
+Similarly, the controller will also the old version to be able to handle
+the fields not present in the new one (and to avoid constant conversion to
+and from the storage version).
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Refactor the BMO code to avoid referring directly to the current version.
+- Add `Architecture` to the current version.
+- Add `empty` hardware profile.
+- Add `auto` checksum type.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Make a new BMO release and communicate the change to the community.
+
+In the future: create a BMO branch and delete `v1alpha1` afterwards.
+This will involve switching the storage format and the controller version
```

Comment:
Do we have to switch these both at the same time, or could there be a period where the v1alpha1 version still exists in the CRD but the storage format is v1beta1 and the controller uses v1beta1? Obviously any v1alpha1-only features would no longer work after this point, but this might be better than ditching the old version.

In fact, I think this may be essential - in order to migrate existing CRs to the new version we first have to switch the storage format and then rewrite each CR. I think this may be impossible if the storage format changes at the same time that the old version disappears.

Reply:
Would you like it to be a lasting state (i.e. something operators may potentially be exposed to)? I guess yes, because they'll need to run storage migration (which is something, as far as I understand, that operators run themselves).

Reply:
That's consistent with my understanding, yes.

Reply:
> could there be a period where the v1alpha1 version still exists in the CRD but the storage format is v1beta1 and the controller uses v1beta1

This is definitely what we should do! The common way to do it is to have a conversion webhook. It can convert everything to the "hub" version (usually the latest, so v1beta1 for us) and it can also convert from the hub version to any other supported version. External tools, humans, etc can use whatever version they want by asking for it as long as it is still supported, but BMO itself would only operate on the hub.

Reply:
@lentzi90 that's not about the conversion webhook, we already have it planned. The problem is: the new version does not have some fields. Once we switch the controller to it, there is no way to handle these fields any more. This will be a feature loss while `v1alpha1` still exists.

Reply:
Understood. Then perhaps v1alpha1 needs to be the hub version so the controller can continue working with that?

Reply:
To add some more context: I think it is required that at least 2 versions are supported at the same time in order to smoothly transition from one to the other. This is how it has been done for cert-manager and Cluster API at least. The thing is, all objects must switch the storage versions and this can only be done when the controller can deal with both versions as far as I know. The steps are like this:

1. BMO with v1alpha1 running. There are BMHs in the cluster obviously stored at v1alpha1.
2. BMO is upgraded to handle v1beta1 *but the existing BMHs are still v1alpha1*. At this point the storage version can be set to v1beta1 and then all BMHs must be "refreshed" in the API to update the stored data.
3. When all objects have been stored at v1beta1 it is safe to drop the v1alpha1 version completely (usually with a new update to the controller and CRDs)

Reply:
> BMO is upgraded to handle v1beta1 but the existing BMHs are still v1alpha1.

Yep, that's what Zane is talking about. At this stage, v1alpha1 still exists, but features that only exist in that old version no longer work.

> all BMHs must be "refreshed" in the API to update the stored data

Do you know how that works, by the way?

Reply:
>>  all BMHs must be "refreshed" in the API to update the stored data
>
> Do you know how that works, by the way?

The storage version is updated on write. So either you wait for everything to get changed by the user or you build a little program that reads out every CR and writes it back again.

Reply:
Exactly! I know CAPI and Cert-manager has both implemented this already so we can take notes from that. They did it through their CLIs though so we may need something different.
Ref for CAPI: https://github.com/kubernetes-sigs/cluster-api/issues/6691#issuecomment-1167607669